crowdsourcehinter/crowdsourcehinter.py

```diff
@@ -0,0 +1,409 @@
+        # there are no hints in hint_database.  this probably will
+        # occur only on the very first run of a unit containing this
+        # block.
+         if not self.hint_database:
+           self.hints_database = self.initial_hints
+
+        answer = self.extract_student_answers(data["submittedanswer"])
+
+        # HACK: For now, we assume just one submission, a string, and
+        # case insensitive
+        #
+        # TODO: We should replace this with a generic canonicalization
+        # function
+        answer = answer.values()[0].lower()
+
+        # Put the student's answer to lower case so that differences
+        # in capitalization don't make different groups of
+        # hints. TODO: We should replace this with a .
+        remaining_hints = 0  # TODO: This is confused
+        best_hint = ""  # TODO: What is this?
+        remaining_hints = self.find_hints(answer)
+        if remaining_hints != 0:
+            for hint in self.hint_database[answer]:
+                if hint not in self.reported_hints.keys():
+                    # if best_hint hasn't been set yet or the rating of hints is greater than the rating of best_hint
+                    if best_hint == "" or self.hint_database[answer][hint] > self.hint_database[answer][best_hint]:
+                        best_hint = hint
+            self.used.append(best_hint)
+            return {'BestHint': best_hint, "StudentAnswer": answer}
+        # find generic hints for the student if no specific hints exist
+        if self.generic_hints:
+            generic_hint = random.choice(self.generic_hints)
+            self.used.append(generic_hint)
+            return {'BestHint': generic_hint, "StudentAnswer": answer}
+        else:
+            # if there are no hints in either the database or generic hints
+            self.used.append("There are no hints for" + " " + answer)
+            return {'BestHint': "Sorry, there are no hints for this answer.", "StudentAnswer": answer}
+
+    def find_hints(self, answer):
+        """
+        This function is used to check that an incorrect answer has
+        available hints to show.  It will also add the incorrect
+        answer test to self.incorrect_answers.
+
+        Args:
+          answer: This is equal to answer from get_hint, the answer the student submitted
+        Returns 0 if no hints to show exist
+
+        """
+        isreported = []
+        self.incorrect_answers.append(answer)
+        if answer not in self.hint_database:
+            # add incorrect answer to hint_database if no precedent exists
+            self.hint_database[answer] = {}
+            return 0
+        for hint_key in self.hint_database[answer]:
+            for reported_keys in self.reported_hints:
+                if hint_key in self.reported_hints:
+                    isreported.append(hint_key)
+        if (len(self.hint_database[answer]) - len(isreported)) > 0:
+            return 1
+        else:
+            return 0
```

Comment:
Should return true or false (or have the int be meaningful -- e.g. # remaining hints)